### ISSUE TITLE
Refactor code base

### ISSUE THREAD

**Issue by HaaLeo** [amaiya/ktrain#191]
# Description
Ktrain is an awesome interface for rapid development of ML applications. However, digging into its source code and trying to understand what is going on is quite challenging, since some files are very big and contain several classes at once. E.g.: [`preprocessor.py`](https://github.com/amaiya/ktrain/blob/master/ktrain/text/preprocessor.py) with >1300 lines of code and [`core.py`](https://github.com/amaiya/ktrain/blob/master/ktrain/core.py). Therefore the learning/understanding curve of the ktrain codebase is steep which may hinders new potential contributors to get started and make ktrain even better.

# Proposal
Break up those overwhelming big entities and put each class in its own file. 

Further considerations to enhance code quality and the development process could be to add
* A linter such as [pylint](https://github.com/PyCQA/pylint)
* Lint and test the code in a CI pipeline. For example [GitHub Actions](https://docs.github.com/en/actions)

I know most of the time refactoring is a "non fun" task. However, it would be greatly appreciated and tremendously help new contributors 🚀 🙂 getting started.

**Comment by amaiya** [comment]
Using CI (and a linter) are good suggestions, and something we will probably examine in future (although I have mixed feelings about pylint).   I'm probably not able to do this any time soon due to time constraints and higher priorities, though. 

In regards to your main proposal of breaking up large files, I find that this is subjective, and there are reasonable arguments on both sides.  I've used popular libraries that have decomposed things to large number of small files and methods that have been criticized for being too hard to follow.  At the same time, there are highly popular ML libraries with a smaller number of module files some of which are nearly 2000 lines of code, yet have lots of users/contributors with no issues or problems with this.  With respect to *ktrain*, I've received some comments in the past about how the code is easier to follow and understand precisely because the code is consolidated in a smaller number of places.  Of course, I'm sure there are also some like you that feel the opposite.  My inclination is to leave this as it is for the time being.  

That said, If you're confused about anything in the code base, feel free to post a question.  Thanks for your comments and interest in the library.
